sqlx-sqlite/src/connection/mod.rs

```diff
@@ -544,3 +549,219 @@ impl Statements {
+        sqlite3_preupdate_new, sqlite3_preupdate_old, sqlite3_value, sqlite3_value_type, SQLITE_OK,
+    };
+    use sqlx_core::error::Error;
+    use std::ffi::CStr;
+    use std::fmt::Debug;
+    use std::os::raw::{c_char, c_int, c_void};
+    use std::panic::catch_unwind;
+    use std::ptr;
+    use std::ptr::NonNull;
+
+    pub struct PreupdateHookResult<'a> {
+        pub operation: SqliteOperation,
+        pub database: &'a str,
+        pub table: &'a str,
+        pub case: PreupdateCase,
+    }
+
+    pub(crate) struct PreupdateHookHandler(
+        NonNull<dyn FnMut(PreupdateHookResult) + Send + 'static>,
+    );
+    unsafe impl Send for PreupdateHookHandler {}
+
+    /// The possible cases for when a PreUpdate Hook gets triggered. Allows access to the relevant
+    /// functions for each case through the contained values.
+    pub enum PreupdateCase {
+        /// Pre-update hook was triggered by an insert.
+        Insert(PreupdateNewValueAccessor),
+        /// Pre-update hook was triggered by a delete.
+        Delete(PreupdateOldValueAccessor),
+        /// Pre-update hook was triggered by an update.
+        Update {
+            old_value_accessor: PreupdateOldValueAccessor,
+            new_value_accessor: PreupdateNewValueAccessor,
+        },
+        /// This variant is not normally produced by SQLite. You may encounter it
+        /// if you're using a different version than what's supported by this library.
+        Unknown,
+    }
+
+    /// An accessor for the old values of the row being deleted/updated during the preupdate callback.
+    #[derive(Debug)]
+    pub struct PreupdateOldValueAccessor {
+        db: *mut sqlite3,
+        old_row_id: i64,
+    }
+
+    impl PreupdateOldValueAccessor {
+        /// Gets the amount of columns in the row being deleted/updated.
+        pub fn get_column_count(&self) -> i32 {
+            unsafe { sqlite3_preupdate_count(self.db) }
+        }
+
+        /// Gets the depth of the query that triggered the preupdate hook.
+        /// Returns 0 if the preupdate callback was invoked as a result of
+        /// a direct insert, update, or delete operation;
+        /// 1 for inserts, updates, or deletes invoked by top-level triggers;
+        /// 2 for changes resulting from triggers called by top-level triggers; and so forth.
+        pub fn get_query_depth(&self) -> i32 {
+            unsafe { sqlite3_preupdate_depth(self.db) }
+        }
+
+        /// Gets the row id of the row being updated/deleted.
+        pub fn get_old_row_id(&self) -> i64 {
+            self.old_row_id
+        }
+
+        /// Gets the value of the row being updated/deleted at the specified index.
+        pub fn get_old_column_value(&self, i: i32) -> Result<SqliteValue, Error> {
+            let mut p_value: *mut sqlite3_value = ptr::null_mut();
+            unsafe {
+                let ret = sqlite3_preupdate_old(self.db, i, &mut p_value);
```

Comment:
This should also check `i` against `sqlite3_preupdate_count` because the result is undefined if the index is out of bounds.

Reply:
> We should handle this as a SqliteValueRef instead, using the same lifetime. The user can then use [.to_owned()](https://github.com/launchbadge/sqlx/blob/1678b19a4672fd6a18b4891c53bf0b57638b92a4/sqlx-sqlite/src/value.rs#L61) to get a fully independent value if they need it.

`SqliteValue::new` calls `sqlite3_value_dup` which creates a copy of the `sqlite3_value`. I tested this out by storing the returned `SqliteValue` in a mutex and ensuring the returned value could still be decoded properly after the callback was completed.

I can add something to `SqliteValueRef` for this to avoid the call to `sqlite3_value_dup`, but I think that would require re-implementing a lot of the logic in `SqliteValue` or modifying it so that it can operate on both an owned or borrowed value. Let me know if I'm missing something.

Reply:
> This should also check i against sqlite3_preupdate_count because the result is undefined if the index is out of bounds.

I think the documentation might be a bit conservative here (or they don't want to make any guarantees) because it does [check for these conditions and return an error](https://github.com/sqlite/sqlite/blob/9ee02515c45b8de7586e01024221cd9eb4d9b40b/src/vdbeapi.c#L2345), but I went ahead and added an explicit check here too.